File: Brute Force/Depth-First Search/graph_edu.cpp
#include "algorithm-visualizer.h"
#include <iostream>
#include <vector>
#include <stack>

const int				GRAPHSIZE = /*(그래프 사이즈 입력)*/
//visualizer{
GraphTracer				graphTracer("GraphTracer");
Array1DTracer			visitedTracer("visited");
LogTracer				logger("LogTracer");
Randomize::Graph<int>	graph(GRAPHSIZE, 0.3);
int						G[GRAPHSIZE][GRAPHSIZE];
int						count = 0;	//무한루프 방지

void InitDepthFirstSearch_graph()
{
	graphTracer.directed(false);
	Layout::setRoot(VerticalLayout({ graphTracer,visitedTracer,logger }));
	graphTracer.log(logger);
	graph.directed(false);
	int temp_G[GRAPHSIZE*GRAPHSIZE];
	graph.fill(temp_G);
	for (int i = 0; i < GRAPHSIZE; i++) {
		for (int j = 0; j < GRAPHSIZE; j++) {
			G[i][j] = temp_G[i * GRAPHSIZE + j];
		}
	}
	graphTracer.set(G);
	Tracer::delay();
}
//}

std::vector<bool> DFSExplore(int graph[GRAPHSIZE][GRAPHSIZE], int source)
{
	//graph: DFS 방법으로 connected인지 확인할 대상이 되는 그래프
	//탐색에 있어서 출발점이 되는 노드
	std::stack<std::array<int, 2>> stack; stack.push({ source, -1 });		
	//stack에 저장되는 데이터는 {현제노드, 부모노드}임 (-1은 노드가 없음을 의미)
	std::vector<bool> visited;	//방문 유무를 담을 컨테이너
	int node, prev, i;
	for (int i = 0; i < GRAPHSIZE; i++) {
		visited.push_back(false);	//초기에는 모든 노드를 미방문으로 세팅
	}
	//visualizer{
	visitedTracer.set(visited);
	//}
	while (/*(알맞은 조건 입력)*/)
	{
		auto temp = stack.top(); stack.pop();
		node = temp[0];
		prev = temp[1];

		if (!visited[node]) {//현재 노드를 방문상태로 표시
			visited[node] = /*(알맞은 값 입력)*/
			//visualizer{
			visitedTracer.patch(node, true);
			//}

			if (prev != -1 && graph[node][prev]) {
				//visualizer{
				graphTracer.visit(node, prev);
				Tracer::delay();
				//}
			}
			else
			{
				//visualizer{
				graphTracer.visit(node);
				Tracer::delay();
				//}
			}

			for (int i = 0; i < GRAPHSIZE; i++) {
				if (graph[node][i]) {
					stack.push({ i,node });
				}
			}
		}
		//visualizer{
		if (count++ > GRAPHSIZE*GRAPHSIZE) break;	//무한루프 방지
		//}
	}
	return visited;
}

int main()
{
	InitDepthFirstSearch_graph();
	std::vector<bool> visited = DFSExplore(G, 0);
	bool check = true;
	for (int i = 0; i < visited.size(); i++) check &= visited[i];
	if (check) {
		//visualizer{
		logger.println("The Graph is CONECTED");
		//}
	}
	else
	{
		//visualizer{
		logger.println("The Graph is NOT CONNECTED");
		//}
	}
}

File: Brute Force/Depth-First Search/graph_edu.java
import org.algorithm_visualizer.*;
import java.util.*;

class NodeData
{
	//각각의 노드는 이전 탐색한 노드의 정보와 현재 노드의 정보를 갖고 있음
	int node;
	int prev;
	public NodeData(int node, int prev)
	{
		this.node = node;
		this.prev = prev;
	}
}

public class Main {
	//visualizer{
	static private GraphTracer graphTracer;
	static private Array1DTracer visitedTracer;
	static private LogTracer logger;
	static private Randomize.Graph G;
	//}
	static private Object[][] graph; //랜덤으로 생성된 graph 2차원 배열
	                                 //배열의 원소값은, 0과 1로 구성
	                                 //graph[i][j] = 1 : i, j 노드가 인접
	                                 //graph[i][j] = 0 : i, j 노드가 인접하지 않음
	                                 
	static private String[] visited; //탐색을 하며, 방문 되었는지 기록을 하는 배열 방문했다면 T, 
	                                 //방문하지 않았다면 F 를 배열에 저장
	static private NodeData temp;    
	static private int number_Of_Node = /*생성할 노드의 개수입력*/
	Main() 
	{
		//visualizer{
		graphTracer = new GraphTracer().directed(false).weighted();
		logger = new LogTracer();
		visitedTracer = new Array1DTracer();
		G = new Randomize.Graph(number_Of_Node, .3);
		Layout.setRoot(new VerticalLayout(new Commander[] {graphTracer,visitedTracer, logger}));
		graphTracer.log(logger);
		G.directed(false).weighted(true);
		graph = G.create();
		graphTracer.set(graph);
		Tracer.delay();
		//}
        
        //함수 실행
		DFSExplore();
		
		boolean check = true;
        
 		for (int i = 0; i < visited.length; i++)
 		{
 			if(visited[i] != /*?*/)
 				check = false;
 		}
 	
		if (/*알맞은 조건*/)
		{
			logger.println("The Graph is CONNECTED");
		}
		else
		{
			logger.println("The Graph is NOT CONNECTED");
		}
	}

	void DFSExplore()
	{
		Stack<NodeData> stack = new Stack<NodeData>() ;
		//처음 방문할 노드를 stack에 ???
		stack./*?*/( new NodeData(0,-1) );
		visited = new String[number_Of_Node];
		
		//visited 배열의 모든 원소를 F로 초기화
		for (int i = 0; i < visited.length; i++)
			visited[i] = "F";
		
		//visualizer{
		visitedTracer.set(visited);
        //}
        
		while (stack.size() > 0)
		{
			temp = stack./*?*/;
			int node = /* ? */;
			int prev = /* ? */;
			
			if (visited[ node ] == "F")
			{
				
				//visualizer{
                Tracer.delay();
                //}
                visited[ node ] = /*?*/;
                //visualizer{
				visitedTracer.patch( node, visited[node]);
                //}
                
				if (prev != -1 && (int)graph[node][prev] != 0)
				{
				    //visualizer{
					graphTracer.visit(node, prev);
					Tracer.delay();
					//}
				}
				else
				{
				    //visualizer{
					graphTracer.visit(node);
					Tracer.delay();
					//}
				}

				for (int i = 0; i < graph.length; i++)
				{
					if ((int)graph[node][i] != /*?*/)
					{
						stack./*?*/(new NodeData( /*?*/, /*?*/ ));
					}
				}
			}
		}
	}

	public static void main(String[] args) {
		new Main();
	}
}